File: src/main/java/com/antlr/plugin/resolve/ImportResolver.java
package com.antlr.plugin.resolve;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiRecursiveElementVisitor;
import com.intellij.psi.util.PsiTreeUtil;
import com.antlr.plugin.parser.ANTLRv4Parser;
import com.antlr.plugin.psi.GrammarElementRefNode;
import com.antlr.plugin.psi.GrammarSpecNode;
import com.antlr.plugin.psi.MyPsiUtils;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static com.antlr.plugin.ANTLRv4TokenTypes.RULE_ELEMENT_TYPES;
import static com.antlr.plugin.resolve.TokenVocabResolver.findRelativeFile;


public class ImportResolver {

    public static PsiFile resolveImportedFile(GrammarElementRefNode reference) {
        PsiElement importStatement = PsiTreeUtil.findFirstParent(reference, ImportResolver::isImportStatement);

        if (importStatement != null) {
            return findRelativeFile(reference.getText(), reference.getContainingFile());
        }

        return null;
    }

    private static boolean isImportStatement(PsiElement el) {
        ASTNode node = el.getNode();
        return node != null && node.getElementType() == RULE_ELEMENT_TYPES.get(ANTLRv4Parser.RULE_delegateGrammar);
    }

    public static PsiElement resolveInImportedFiles(@NotNull PsiFile grammarFile, @NotNull String ruleName) {
        return resolveInImportedFiles(grammarFile, ruleName, new ArrayList<>());
    }

    private static PsiElement resolveInImportedFiles(PsiFile grammarFile, String ruleName, List<PsiFile> visitedFiles) {
        DelegateGrammarsVisitor visitor = new DelegateGrammarsVisitor();
        grammarFile.accept(visitor);

        for (PsiFile importedGrammar : visitor.importedGrammars) {
            if (visitedFiles.contains(importedGrammar)) {
                continue;
            }
            visitedFiles.add(importedGrammar);

            GrammarSpecNode grammar = PsiTreeUtil.getChildOfType(importedGrammar, GrammarSpecNode.class);
            PsiElement specNode = MyPsiUtils.findSpecNode(grammar, ruleName);

            if (specNode != null) {
                return specNode;
            }

            // maybe the imported grammar also imports other grammars itself?
            specNode = resolveInImportedFiles(importedGrammar, ruleName, visitedFiles);
            if (specNode != null) {
                return specNode;
            }
        }

        return null;
    }

    private static class DelegateGrammarsVisitor extends PsiRecursiveElementVisitor {

        List<PsiFile> importedGrammars = new ArrayList<>();

        @Override
        public void visitElement(@NotNull PsiElement element) {
            if (isImportStatement(element)) {
                PsiFile importedGrammar = findRelativeFile(element.getText(), element.getContainingFile());

                if (importedGrammar != null) {
                    importedGrammars.add(importedGrammar);
                }
            }
            super.visitElement(element);
        }
    }
}

File: src/main/java/com/antlr/language/resolve/ImportResolver.kt
package com.antlr.language.resolve

import com.antlr.language.ANTLRv4Parser
import com.antlr.language.AntlrTokenTypes.RULE_ELEMENT_TYPES
import com.antlr.language.psi.GrammarElementRefNode
import com.antlr.language.psi.GrammarSpecNode
import com.antlr.language.psi.MyPsiUtils.findSpecNode
import com.intellij.openapi.util.Condition
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiRecursiveElementVisitor
import com.intellij.psi.util.PsiTreeUtil

object ImportResolver {
    fun resolveImportedFile(reference: GrammarElementRefNode?): PsiFile? {
        val importStatement =
            PsiTreeUtil.findFirstParent(reference, ImportResolver::isImportStatement)

        if (importStatement != null) {
            return TokenVocabResolver.findRelativeFile(reference!!.text, reference.containingFile)
        }

        return null
    }

    private fun isImportStatement(el: PsiElement): Boolean {
        val node = el.node
        return node != null && node.elementType === RULE_ELEMENT_TYPES?.get(ANTLRv4Parser.RULE_delegateGrammar)
    }

    fun resolveInImportedFiles(grammarFile: PsiFile, ruleName: String): PsiElement? {
        return resolveInImportedFiles(grammarFile, ruleName, ArrayList<PsiFile?>())
    }

    private fun resolveInImportedFiles(
        grammarFile: PsiFile,
        ruleName: String?,
        visitedFiles: MutableList<PsiFile?>
    ): PsiElement? {
        val visitor = DelegateGrammarsVisitor()
        grammarFile.accept(visitor)

        for (importedGrammar in visitor.importedGrammars) {
            if (visitedFiles.contains(importedGrammar)) {
                continue
            }
            visitedFiles.add(importedGrammar)

            val grammar = PsiTreeUtil.getChildOfType<GrammarSpecNode?>(importedGrammar, GrammarSpecNode::class.java)
            var specNode = findSpecNode(grammar, ruleName)

            if (specNode != null) {
                return specNode
            }

            // maybe the imported grammar also imports other grammars itself?
            specNode = resolveInImportedFiles(importedGrammar, ruleName, visitedFiles)
            if (specNode != null) {
                return specNode
            }
        }

        return null
    }

    private class DelegateGrammarsVisitor : PsiRecursiveElementVisitor() {
        var importedGrammars: MutableList<PsiFile> = ArrayList<PsiFile>()

        override fun visitElement(element: PsiElement) {
            if (isImportStatement(element)) {
                val importedGrammar =
                    TokenVocabResolver.findRelativeFile(element.text, element.containingFile)

                if (importedGrammar != null) {
                    importedGrammars.add(importedGrammar)
                }
            }
            super.visitElement(element)
        }
    }
}
